dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java

```diff
@@ -658,4 +684,61 @@ private String gitHubAppSetup(String installationId) {
         }
         return installationAccessToken;
     }
+
+    /**
+     * Publish or unpublish given workflow
+     * @param workflow
+     * @param publish
+     * @return
+     */
+    Workflow publishWorkflow(Workflow workflow, final boolean publish) {
+        Workflow checker = workflow.getCheckerWorkflow();
+
+        if (workflow.isIsChecker()) {
+            String msg = "Cannot directly publish/unpublish a checker workflow.";
+            LOG.error(msg);
+            throw new CustomWebApplicationException(msg, HttpStatus.SC_BAD_REQUEST);
+        }
+
+        if (publish) {
+            boolean validTag = false;
+            Set<WorkflowVersion> versions = workflow.getWorkflowVersions();
+            for (WorkflowVersion workflowVersion : versions) {
+                if (workflowVersion.isValid()) {
+                    validTag = true;
+                    break;
+                }
+            }
+
+            if (validTag && (!workflow.getGitUrl().isEmpty() || Objects.equals(workflow.getMode(), WorkflowMode.HOSTED))) {
+                workflow.setIsPublished(true);
+                if (checker != null) {
+                    checker.setIsPublished(true);
+                }
+            } else {
+                throw new CustomWebApplicationException("Repository does not meet requirements to publish.", HttpStatus.SC_BAD_REQUEST);
+            }
+        } else {
+            workflow.setIsPublished(false);
+            if (checker != null) {
+                checker.setIsPublished(false);
+            }
+        }
+
+        long id = workflowDAO.create(workflow);
+        workflow = workflowDAO.findById(id);
+        if (publish) {
+            PublicStateManager.getInstance().handleIndexUpdate(workflow, StateManagerMode.PUBLISH);
+            if (workflow.getTopicId() == null) {
+                try {
+                    entryResource.createAndSetDiscourseTopic(id);
+                } catch (CustomWebApplicationException ex) {
+                    LOG.error("Error adding discourse topic.", ex);
+                }
+            }
+        } else {
+            PublicStateManager.getInstance().handleIndexUpdate(workflow, StateManagerMode.DELETE);
```

Comment:
Need to understand, why is this always creating but sometimes deleting. Might tie in with previous question

Reply:
This is the existing publish code transferred from `WorkflowResource` without deviation.
https://github.com/dockstore/dockstore/blob/9c48baeed4d037da4d5d66688e77a32cb49e5064/dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java#L874
Can't say I understand why.

Reply:
Please debug the control flow , especially `createOrGetWorkflow` 
What I'm puzzled about it is looks like  `workflowDAO.create` might be getting called once there and once here. 
I can understand if `workflow` is the deserialized workflow coming from a web request, then we always need to make a DB copy whether published or not. But if it is actually the workflow coming from the webservice then I'm a little confused and we might be creating an extra workflow by accident.

Reply:
I removed the seemingly redundant `workflowDAO.create` in the publish code and it seemed to not break anything, so we can stick with that for now.